File: Heat/Claims/HttpModules/HeatClaimsBasedAuthorization.vb
Imports System.Security.Claims

Public Class HeatClaimsBasedAuthorization
    Implements IHttpModule

    Public Sub Dispose() Implements IHttpModule.Dispose

    End Sub

    Public Sub Init(context As HttpApplication) Implements IHttpModule.Init
        AddHandler context.PostAuthenticateRequest, AddressOf ReplacePrincipal
    End Sub
    Private Shared Sub ReplacePrincipal()
        If Not IsNothing(HttpContext.Current.User) AndAlso HttpContext.Current.User.Identity.IsAuthenticated Then
            Dim cp As ClaimsPrincipal = CreateClaimsBasedPrincipal()
        End If

    End Sub

    Private Shared Function CreateClaimsBasedPrincipal() As ClaimsPrincipal
        Dim UserName As String = Threading.Thread.CurrentPrincipal.Identity.Name
        Dim hcp As HeatClaimsPrincipal = New HeatClaimsPrincipal(UserName)
        Threading.Thread.CurrentPrincipal = hcp

        If Not IsNothing(HttpContext.Current) Then
            HttpContext.Current.User = hcp
        End If

        Return hcp
    End Function

    
End Class

File: Heat.ConvertedToC#/Claims/HttpModules/HeatClaimsBasedAuthorization.cs
using System;
using System.Web;
using System.Security.Claims;
namespace Heat
{

    public class HeatClaimsBasedAuthorization : IHttpModule
	{


		public void Dispose()
		{
		}

		public void Init(HttpApplication context)
		{
			context.PostAuthenticateRequest += ReplacePrincipal;
		}
		private static void ReplacePrincipal(object sender, EventArgs e)
		{
			if ((HttpContext.Current.User != null) && HttpContext.Current.User.Identity.IsAuthenticated) {
				ClaimsPrincipal cp = CreateClaimsBasedPrincipal();
			}

		}

		private static ClaimsPrincipal CreateClaimsBasedPrincipal()
		{
			string UserName = System.Threading.Thread.CurrentPrincipal.Identity.Name;
			HeatClaimsPrincipal hcp = new HeatClaimsPrincipal(UserName);
			System.Threading.Thread.CurrentPrincipal = hcp;

			if ((HttpContext.Current != null)) {
				HttpContext.Current.User = hcp;
			}

			return hcp;
		}


	}
}
